consensus/src/transaction_shuffler/fairness/mod.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transaction_shuffler::{
+    fairness::{
+        conflict_key::ConflictKeyRegistry, conflict_zone::ConflictZone, pending_zone::PendingZone,
+    },
+    TransactionShuffler,
+};
+use aptos_types::transaction::SignedTransaction;
+use itertools::zip_eq;
+use selection_tracker::SelectionTracker;
+use std::collections::BTreeSet;
+
+pub(crate) mod conflict_key;
+mod conflict_zone;
+mod pending_zone;
+mod selection_tracker;
+
+#[cfg(test)]
+mod tests;
+
+type TxnIdx = usize;
+
+const NUM_CONFLICT_ZONES: usize = 3;
+
+#[derive(Debug)]
+struct FairnessShuffler {
+    sender_conflict_window_size: usize,
+    module_conflict_window_size: usize,
+    entry_fun_conflict_window_size: usize,
+}
+
+impl TransactionShuffler for FairnessShuffler {
+    fn shuffle(&self, txns: Vec<SignedTransaction>) -> Vec<SignedTransaction> {
```

Comment:
We would likely need some onchain configuration changes to enable this, I don't see that in this PR. Is that a follow up?

Reply:
yeah for sure (this is still draft)